### PR TITLE
General Filesystem and Save Data Fixes

### PR DESCRIPTION
**Pull request by DarkLordZach**
1. Save data now uses the following directory structure:
`%YUZU_DIR&/save/{storage_space}/{type}/{title_id}/{user_id}/{save_id}/.`
Where:
- `storage_space` is `sysnand`, `usrnand`, `sd`, or `temp`
- `type` is `system`, `user`, `bcat`, `device`, `temp`, or `cache`
- `user_id` is the uid of the current user (maybe multiple users in future [no promises], for right now defaults to `00 00 00 00 00 00 00 00 00 00 00 00 10 00 00 00`)

2. The save data code now supports multiple users, save ids, different types and different title ids. (i.e. no longer assuming the current process' title id).

Most of this might be extra, but is here for correctness.

### CODE REVIEW

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 1]
```diff
@@ -16,57 +16,91 @@ namespace Service::FileSystem {
...
 
-ResultVal<std::unique_ptr<FileSys::FileSystemBackend>> OpenFileSystem(Type type,
-                                                                      FileSys::Path& path) {
-    LOG_TRACE(Service_FS, "Opening FileSystem with type={}", static_cast<u32>(type));
+ResultVal<std::unique_ptr<FileSys::FileSystemBackend>> OpenRomFS(u64 title_id) {
+    LOG_TRACE(Service_FS, "Opening RomFS for title_id={:016X}", title_id);
+
+    if (romfs == nullptr) {
+        // TODO(bunnei): Find a better error code for this
+        return ResultCode(-1);
```
Comment (ogniK5377) [inline src/core/hle/service/filesystem/filesystem.cpp]: We could use 0x2EE202(`ResultCode(ErrorModule::FS, 6001);`) which is Invalid Input or we can use 0x1F4202(`ResuleCode(ErrorModue::FS, 4001);`) Which is corrupted RomFS(not sure if used in retail FS though).

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 2]
```diff
@@ -16,57 +16,91 @@ namespace Service::FileSystem {
...
-    LOG_TRACE(Service_FS, "Formatting FileSystem with type={}", static_cast<u32>(type));
+ResultCode FormatSaveData(FileSys::SaveDataSpaceId space, FileSys::SaveStruct save_struct) {
+    LOG_TRACE(Service_FS, "Formatting Save Data for space_id={:01X}, save_struct={}",
+              static_cast<u8>(space), SaveStructDebugInfo(save_struct));
 
-    auto itr = filesystem_map.find(type);
-    if (itr == filesystem_map.end()) {
+    if (save_data == nullptr) {
         // TODO(bunnei): Find a better error code for this
         return ResultCode(-1);
```
Comment (ogniK5377) [inline src/core/hle/service/filesystem/filesystem.cpp]: Should return 0x7D402 `ResuleCode(ErrorModue::FS, 1002);`

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 3]
```diff
@@ -16,57 +16,91 @@ namespace Service::FileSystem {
...
+}
+
+ResultVal<std::unique_ptr<FileSys::FileSystemBackend>> OpenSDMC() {
+    LOG_TRACE(Service_FS, "Opening SDMC");
 
-    auto itr = filesystem_map.find(type);
-    if (itr == filesystem_map.end()) {
+    if (sdmc == nullptr) {
         // TODO(bunnei): Find a better error code for this
         return ResultCode(-1);
```
Comment (ogniK5377) [inline src/core/hle/service/filesystem/filesystem.cpp]: Should be 0xFA202 `ResuleCode(ErrorModue::FS, 2001);`

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 4]
```diff
@@ -583,8 +585,35 @@ void FSP_SRV::OpenDataStorageByCurrentProcess(Kernel::HLERequestContext& ctx) {
...
+
+    LOG_DEBUG(Service_FS, "called with storage_id={:02X}, title_id={:016X}",
+              static_cast<u8>(storage_id), title_id);
+
+    auto romfs = OpenRomFS(title_id);
+    if (romfs.Failed()) {
+        // TODO (bunnei): Find the right error code to use here
+        LOG_CRITICAL(Service_FS, "no file system interface available!");
+        IPC::ResponseBuilder rb{ctx, 2};
+        rb.Push(ResultCode(-1));
```
Comment (ogniK5377) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: 0x13B002 might be best used here which is "Gamecard not inserted"
`ResultCode(ErrorModule::FS, 2520);`

File: src/core/hle/service/filesystem/filesystem.cpp (hunk 1 above)
Comment (DarkLordZach) [reply src/core/hle/service/filesystem/filesystem.cpp]: I'm planning on the 6001, (seems more reasonable). Should i leave a todo?

Comment (Subv) [reply src/core/hle/service/filesystem/filesystem.cpp]: If the error code has not been verified by a hardware test, leave -1 and the todo.

File: src/core/file_sys/savedata_factory.cpp [hunk 5]
```diff
@@ -12,11 +12,50 @@
...
+                       static_cast<u8>(save_struct.type), save_struct.title_id,
+                       save_struct.user_id[1], save_struct.user_id[0], save_struct.save_id);
+}
+
+SaveDataFactory::SaveDataFactory(std::string save_directory)
+    : save_directory(std::move(save_directory)) {}
+
+ResultVal<std::unique_ptr<FileSystemBackend>> SaveDataFactory::Open(SaveDataSpaceId space,
+                                                                    SaveStruct meta) {
+    if ((meta.type == SaveDataType::SystemSaveData || meta.type == SaveDataType::SaveData)) {
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.cpp]: why the double parens?

File: src/core/file_sys/savedata_factory.cpp [hunk 6]
```diff
@@ -36,28 +81,48 @@ ResultVal<std::unique_ptr<FileSystemBackend>> SaveData_Factory::Open(const Path&
     return MakeResult<std::unique_ptr<FileSystemBackend>>(std::move(archive));
 }
 
-ResultCode SaveData_Factory::Format(const Path& path) {
-    LOG_WARNING(Service_FS, "Format archive {}", GetName());
+ResultCode SaveDataFactory::Format(SaveDataSpaceId space, SaveStruct meta) {
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.cpp]: Remove this method

File: src/core/file_sys/savedata_factory.h [hunk 7]
```diff
@@ -12,22 +12,44 @@
 
 namespace FileSys {
 
+enum class SaveDataSpaceId : u8 { NandSystem = 0, NandUser = 1, SdCard = 2, TemporaryStorage = 3 };
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.h]: add a `,` to the last element.

File: src/core/file_sys/savedata_factory.h [hunk 8]
```diff
@@ -12,22 +12,44 @@
...
+
+enum class SaveDataType : u8 {
+    SystemSaveData = 0,
+    SaveData = 1,
+    BcatDeliveryCacheStorage = 2,
+    DeviceSaveData = 3,
+    TemporaryStorage = 4,
+    CacheStorage = 5
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.h]: same

File: src/core/file_sys/savedata_factory.h [hunk 9]
```diff
@@ -12,22 +12,44 @@
...
+};
+
+struct SaveStruct {
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.h]: Change the name of this to something more descriptive, maybe SaveDataPath

File: src/core/hle/service/am/am.cpp [hunk 10]
```diff
@@ -614,25 +616,45 @@ void IApplicationFunctions::CreateApplicationAndRequestToStartForQuest(
...
 void IApplicationFunctions::EnsureSaveData(Kernel::HLERequestContext& ctx) {
     IPC::RequestParser rp{ctx};
-    u128 uid = rp.PopRaw<u128>();
+    u128 uid = rp.PopRaw<u128>(); // What does this do?
+
+    // TODO(DarkLordZach): So the next bit is a huge assumption. I don't think this method is
+    // supposed to do this, as there is no savestruct provided.
+
+    FileSys::SaveDataSpaceId space = FileSys::SaveDataSpaceId::NandUser; // Default space
+    FileSys::SaveStruct save_struct = {
```
Comment (Subv) [inline src/core/hle/service/am/am.cpp]: Just stub this entire function, remove the code in here. The save will be created upon trying to open it anyways. We need to investigate this more.

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 11]
```diff
@@ -16,57 +16,91 @@ namespace Service::FileSystem {
  * Map of registered file systems, identified by type. Once an file system is registered here, it
  * is never removed until UnregisterFileSystems is called.
  */
-static boost::container::flat_map<Type, std::unique_ptr<FileSys::FileSystemFactory>> filesystem_map;
-
-ResultCode RegisterFileSystem(std::unique_ptr<FileSys::FileSystemFactory>&& factory, Type type) {
-    auto result = filesystem_map.emplace(type, std::move(factory));
+static std::unique_ptr<FileSys::RomFSFactory> romfs;
```
Comment (Subv) [inline src/core/hle/service/filesystem/filesystem.cpp]: add _factory to these variables

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 12]
```diff
@@ -16,57 +16,91 @@ namespace Service::FileSystem {
...
         return ResultCode(-1);
     }
 
-    return itr->second->Open(path);
+    return sdmc->Open();
 }
 
-ResultCode FormatFileSystem(Type type) {
-    LOG_TRACE(Service_FS, "Formatting FileSystem with type={}", static_cast<u32>(type));
+ResultCode FormatSaveData(FileSys::SaveDataSpaceId space, FileSys::SaveStruct save_struct) {
```
Comment (Subv) [inline src/core/hle/service/filesystem/filesystem.cpp]: remove this function

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 13]
```diff
@@ -16,57 +16,91 @@ namespace Service::FileSystem {
...
+    sdmc = nullptr;
 
-    std::string nand_directory = FileUtil::GetUserPath(D_NAND_IDX);
+    std::string save_directory = FileUtil::GetUserPath(D_SAVE_IDX);
     std::string sd_directory = FileUtil::GetUserPath(D_SDMC_IDX);
 
-    auto savedata = std::make_unique<FileSys::SaveData_Factory>(std::move(nand_directory));
-    RegisterFileSystem(std::move(savedata), Type::SaveData);
+    auto savedata = std::make_unique<FileSys::SaveDataFactory>(std::move(save_directory));
+    RegisterSaveData(std::move(savedata));
```
Comment (Subv) [inline src/core/hle/service/filesystem/filesystem.cpp]: you can remove the Register* functions and just assign the variables directly here

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 14]
```diff
@@ -2,6 +2,8 @@
 // Licensed under GPLv2 or any later version
 // Refer to the license.txt file included.
 
+#pragma optimize("", off)
```
Comment (Subv) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: remove this

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 15]
```diff
@@ -509,7 +510,7 @@ void FSP_SRV::MountSdCard(Kernel::HLERequestContext& ctx) {
     LOG_DEBUG(Service_FS, "called");
 
     FileSys::Path unused;
-    auto filesystem = OpenFileSystem(Type::SDMC, unused).Unwrap();
+    auto filesystem = IFileSystem(OpenSDMC().Unwrap());
```
Comment (Subv) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: IFileSystem filesystem{OpenSDMC().Unwrap());

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 16]
```diff
@@ -519,23 +520,24 @@ void FSP_SRV::MountSdCard(Kernel::HLERequestContext& ctx) {
...
-    LOG_WARNING(Service_FS, "(STUBBED) called");
+    IPC::RequestParser rp{ctx};
 
-    // TODO(Subv): Read the input parameters and mount the requested savedata instead of always
-    // mounting the current process' savedata.
-    FileSys::Path unused;
-    auto filesystem = OpenFileSystem(Type::SaveData, unused);
+    auto space_id = rp.PopRaw<FileSys::SaveDataSpaceId>();
+    auto save_struct = rp.PopRaw<FileSys::SaveStruct>();
+    save_struct.type = FileSys::SaveDataType::SaveData;
```
Comment (Subv) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: do you have to set this yourself?

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 17]
```diff
@@ -519,23 +520,24 @@ void FSP_SRV::MountSdCard(Kernel::HLERequestContext& ctx) {
...
+    auto filesystem = OpenSaveData(static_cast<FileSys::SaveDataSpaceId>(space_id), save_struct);
```
Comment (Subv) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: Is this not already a FileSys::SaveDataSpaceId?

File: src/core/hle/service/filesystem/fsp_srv.cpp [hunk 18]
```diff
@@ -583,8 +585,35 @@ void FSP_SRV::OpenDataStorageByCurrentProcess(Kernel::HLERequestContext& ctx) {
...
-    OpenDataStorageByCurrentProcess(ctx);
+    IPC::RequestParser rp{ctx};
+
+    auto storage_id = rp.PopRaw<StorageId>();
+    auto title_id = rp.PopRaw<u64>();
+
+    LOG_DEBUG(Service_FS, "called with storage_id={:02X}, title_id={:016X}",
+              static_cast<u8>(storage_id), title_id);
+
+    auto romfs = OpenRomFS(title_id);
```
Comment (Subv) [inline src/core/hle/service/filesystem/fsp_srv.cpp]: assert that the title_id is the same as the current process'

File: src/core/file_sys/savedata_factory.cpp [hunk 19]
```diff
@@ -113,16 +113,31 @@ ResultCode SaveDataFactory::Format(SaveDataSpaceId space, SaveStruct meta) {
...
-                       title_id, user_id[1], user_id[0], save_id);
+    std::string prefix;
+
+    switch (space) {
+    case SaveDataSpaceId::NandSystem:
+        prefix = nand_directory + "system/save/";
+    case SaveDataSpaceId::NandUser:
+        prefix = nand_directory + "user/save/";
+    default:
+        ASSERT_MSG(true, "Unrecognized SaveDataSpaceId: {:02X}", static_cast<u8>(space));
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.cpp]: ASSERT_MSG(false, ...)

File: src/core/file_sys/savedata_factory.cpp [hunk 20]
```diff
@@ -113,16 +113,31 @@ ResultCode SaveDataFactory::Format(SaveDataSpaceId space, SaveStruct meta) {
...
+    }
+
+    switch (type) {
+    case SaveDataType::SystemSaveData:
+        return fmt::format("{}{:016X}/{:016X}{:016X}", prefix, save_id, user_id[1], user_id[0]);
+    case SaveDataType::SaveData:
+        return fmt::format("{}{:08X}/{:016X}{:016X}/{:016X}", prefix, 0, user_id[1], user_id[0],
+                           title_id);
+    default:
+        ASSERT_MSG(true, "Unrecognized SaveDataType: {:02X}", static_cast<u8>(type));
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.cpp]: ASSERT_MSG(false, ...)

File: src/core/file_sys/savedata_factory.cpp [hunk 21]
```diff
@@ -113,16 +113,31 @@ ResultCode SaveDataFactory::Format(SaveDataSpaceId space, SaveStruct meta) {
...
+        prefix = nand_directory + "user/save/";
+    default:
+        ASSERT_MSG(true, "Unrecognized SaveDataSpaceId: {:02X}", static_cast<u8>(space));
+    }
+
+    switch (type) {
+    case SaveDataType::SystemSaveData:
+        return fmt::format("{}{:016X}/{:016X}{:016X}", prefix, save_id, user_id[1], user_id[0]);
+    case SaveDataType::SaveData:
+        return fmt::format("{}{:08X}/{:016X}{:016X}/{:016X}", prefix, 0, user_id[1], user_id[0],
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.cpp]: Make that 16 zeroes instead of 8

File: src/core/file_sys/savedata_factory.h [hunk 22]
```diff
@@ -38,15 +38,17 @@ static_assert(sizeof(SaveStruct) == 0x40, "SaveStruct has incorrect size.");
 std::string SaveStructDebugInfo(SaveStruct save_struct);
 
 /// File system interface to the SaveData archive
+// nand_directory should be the root of nand, sd_directory should be the root of sd
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.h]: This comment is redundant

File: src/core/file_sys/savedata_factory.h [hunk 23]
```diff
@@ -38,15 +38,17 @@ static_assert(sizeof(SaveStruct) == 0x40, "SaveStruct has incorrect size.");
...
 class SaveDataFactory {
 public:
-    explicit SaveDataFactory(std::string save_directory);
+    SaveDataFactory(std::string nand_directory);
```
Comment (Subv) [inline src/core/file_sys/savedata_factory.h]: Why did you remove the explicit?

File: src/core/file_sys/savedata_factory.cpp [hunk 24]
```diff
@@ -12,11 +12,50 @@
 
 namespace FileSys {
 
-SaveData_Factory::SaveData_Factory(std::string nand_directory)
+std::string SaveStructDebugInfo(SaveDataDescriptor save_struct) {
```
Comment (bunnei) [inline src/core/file_sys/savedata_factory.cpp]: nit: This could just be a const method on `SaveDataDescriptor`

File: src/core/file_sys/savedata_factory.cpp [hunk 25]
```diff
@@ -26,6 +65,12 @@ ResultVal<std::unique_ptr<FileSystemBackend>> SaveData_Factory::Open(const Path&
         FileUtil::CreateFullPath(save_directory);
     }
 
+    // TODO(DarkLordZach): For some reason, CreateFullPath dosen't create the last bit. Should be
```
Comment (bunnei) [inline src/core/file_sys/savedata_factory.cpp]: nit: `doesn't`

File: src/core/file_sys/savedata_factory.cpp [hunk 26]
```diff
@@ -36,28 +81,33 @@ ResultVal<std::unique_ptr<FileSystemBackend>> SaveData_Factory::Open(const Path&
...
-ResultCode SaveData_Factory::Format(const Path& path) {
-    LOG_WARNING(Service_FS, "Format archive {}", GetName());
-    // Create the save data directory.
-    if (!FileUtil::CreateFullPath(GetFullPath())) {
-        // TODO(Subv): Find the correct error code.
-        return ResultCode(-1);
-    }
+std::string SaveDataFactory::GetFullPath(SaveDataSpaceId space, SaveDataType type, u64 title_id,
+                                         u128 user_id, u64 save_id) const {
+    if (type == SaveDataType::SaveData && title_id == 0)
```
Comment (bunnei) [inline src/core/file_sys/savedata_factory.cpp]: Whats with this check? Add a comment explaining why this is just for SaveData?

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 27]
```diff
@@ -16,57 +16,77 @@ namespace Service::FileSystem {
  * Map of registered file systems, identified by type. Once an file system is registered here, it
  * is never removed until UnregisterFileSystems is called.
  */
-static boost::container::flat_map<Type, std::unique_ptr<FileSys::FileSystemFactory>> filesystem_map;
-
-ResultCode RegisterFileSystem(std::unique_ptr<FileSys::FileSystemFactory>&& factory, Type type) {
-    auto result = filesystem_map.emplace(type, std::move(factory));
+static std::unique_ptr<FileSys::RomFSFactory> romfs_factory;
```
Comment (bunnei) [inline src/core/hle/service/filesystem/filesystem.cpp]: Ideally, these would be part of Core state, but that is not that important rn

File: src/core/hle/service/filesystem/filesystem.cpp [hunk 28]
```diff
@@ -16,57 +16,77 @@ namespace Service::FileSystem {
...
+    }
+
+    return save_data_factory->Open(space, save_struct);
+}
+
+ResultVal<std::unique_ptr<FileSys::FileSystemBackend>> OpenSDMC() {
+    LOG_TRACE(Service_FS, "Opening SDMC");
+
+    if (sdmc_factory == nullptr) {
+        return ResultCode(ErrorModule::FS, 2001);
```
Comment (Subv) [inline src/core/hle/service/filesystem/filesystem.cpp]: Don't use magic numbers, add these to the ErrCodes namespace of FileSys
